accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,354 @@
+        mi1.InvokeDirect(target: default, arg1: tr);
+
+        // and supported passing byref
+        MethodInfo mi2 = typeof(Program).GetMethod(nameof(ChangeIt2));
+        mi2.InvokeDirect(target: default, arg1: tr);
+
+        // Just like these methods can be called today:
+        ChangeIt1(span);
+        ChangeIt2(ref span);
+    }
+
+    public static void ChangeIt1(Span<int> span) { }
+    public static void ChangeIt2(ref Span<int> span) { }
+}
+```
+
+FWIW `__arglist` currently compiles with byref-like types:
+```cs
+Span<byte> s1 = stackalloc byte[1];
+Span<byte> s2 = stackalloc byte[11];
+Span<byte> s3 = stackalloc byte[111];
+CallMe(__arglist(s1, s2, s3));
+
+static unsafe void CallMe(__arglist)
+{
+    // However, when enumerating __arglist here, the runtime throws when accessing a byref-like
+    // type although that limitation could be removed (just a runtime limitation; not compiler)
+```
+
+## Variable-length, safe collections
+(todo; see https://github.com/dotnet/runtime/issues/75349. _I have a local prototype that does a callback that doesn't require any new language or runtime features however it is not a clean programming model. The callback is necessary to interop with hide our internal GC "tracking" feature via "RegisterForGCReporting()"._)
+
+## Exception handling
+Todo; discuss using `BindingFlags.DoNotWrapExceptions` semantics only (no wrapping of exceptions).
+
+# Proposed APIs
+(work in progress; various prototypes exist here)
+## TypedReference
+```diff
+namespace System
+{
+    public ref struct TypedReference
+    {
+        // Equivalent of __makeref (except for a byref-like type since they can't be a generic parameter)
++       public static TypedReference Make<T>(ref T? value);
+        // Helper used for boxed or loosely-typed cases
++       public static TypedReference Make(ref object value, Type type);
+
+        // Equivalent of __refvalue
++       public ref T GetValue<T>();
+        // Used for byref-like types or loosely-typed cases with >4 parameters
++       public readonly unsafe ref byte TargetRef { get; }
+    }
+}
+```
+
+## MethodInfo
+(these may be extension methods instead)
+
+```diff
+namespace System.Reflection
+{
+    public abstract class MethodBase
+    {
+        // Helpers for <= 4 parameters
+
+        // Note that 'default(TypedReference)' can be specified for any "not used" arguments, such as for
+        // static methods ('target' is 'default'), 'void'-returning methods ('result' is 'default') or
+        // for any trailing unused arguments (e.g. 'arg4' can be 'default').
+
+        // We may also want to consider using 'ref byte' instead of or in addition to 'TypedReference'
+        // along with new helper methods like 'public static ref byte GetRawData(object o)'
++       [System.CLSCompliantAttribute(false)]
++       public virtual void InvokeDirect(
++           TypedReference target,
++           TypedReference result);
```

Comment:
Since this is a bigger struct, it's going to be passed by reference on many platforms anyway. Should we make it explicit?

```suggestion
+           in TypedReference target,
+           out TypedReference result);
```

Reply:
`TypedReference` is same size as `Span<T>` (2*IntPtr.Size). We do not pass `Span<T>` via `in` and it would be de-optimization in typical case. Also, the .NET Framework Design Guidelines are advising against `in` in public APIs (look for "DO NOT pass value types by read-only reference (in).").

> `out TypedReference result`

This would have to be `in TypedReference result`. The `TypedReference` is meant to point to a location to store the result.

Reply:
FWIW Roslyn actually doesn't allow `TypedReference` to be passed today via `in\ref\out` -- CS1601. I don't think we need to change this. Also that would only work if bound to a ref field in a ref struct since it is be stack-allocated.